File: verilog/dv/cocotb/all_tests/user_project/user_address_space.c
#include <firmware_apis.h>



// --------------------------------------------------------

void main()
{
    User_enableIF();
    // first 2 addresses 
    (*(volatile unsigned int*) (USER_SPACE_ADDR    )) = 0x0314DFE1; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR +0x4 )) = 0x3704E836; 

    // last 2 addresses 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + USER_SPACE_SIZE -0x4 )) = 0x5208E431; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + USER_SPACE_SIZE)) = 0x77748E32; 


    // random addresses inside the user space 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x344F4)) = 0x89158A64; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x4F750)) = 0xAE603480; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x6CE0C)) = 0xD24B086A; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x8FEE8)) = 0xE50B1442; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xB767C)) = 0xB23D7EFD; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xD0A00)) = 0x41C35871; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xF11D0)) = 0xC0E1638A; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xC5E54)) = 0x8E16CDA9; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xE9028)) = 0x42EB0C85; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x023BC)) = 0xBF9E7B2E;
    // random read 
    int temp;
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR )); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x4)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + USER_SPACE_SIZE - 0x4)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + USER_SPACE_SIZE)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x344F4)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x4F750)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x6CE0C)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x8FEE8)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xB767C)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xD0A00)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xF11D0)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xC5E54)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0xE9028)); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x023BC)); 

    // addresses outside user space - injecting error if user project ack is affected
    GPIOs_configure(14,GPIO_MODE_MGMT_STD_OUTPUT);
    GPIOs_configure(15,GPIO_MODE_MGMT_STD_OUTPUT);
    
    // finish with writing last address with Fs
    (*(volatile unsigned int*)(USER_SPACE_ADDR + USER_SPACE_SIZE)) = 0xFFFFFFFF; 
    

}

File: verilog/dv/cocotb/all_tests/user_project/user_address_space.py
import cocotb
from cocotb.triggers import RisingEdge, NextTimeStep
import cocotb.log
from caravel_cocotb.caravel_interfaces import test_configure
from caravel_cocotb.caravel_interfaces import report_test
from user_design import configure_userdesign


@cocotb.test()
@report_test
async def user_address_space(dut):
    caravelEnv = await test_configure(dut, timeout_cycles=31776)
    cocotb.log.info("[TEST] Start user_address_space test")
    ack_hdl = caravelEnv.caravel_hdl.mprj.wbs_ack_o
    addr_hdl = caravelEnv.caravel_hdl.mprj.wbs_adr_i
    data_o_hdl = caravelEnv.caravel_hdl.mprj.wbs_dat_o
    data_i_hdl = caravelEnv.caravel_hdl.mprj.wbs_dat_i
    we_hdl = caravelEnv.caravel_hdl.mprj.wbs_we_i
    
    start_addr = int(caravelEnv.design_macros.USER_SPACE_ADDR)
    print(f"user space adddress = {start_addr}")
    user_size = int(caravelEnv.design_macros.USER_SPACE_SIZE)
    addr_arr = (
        start_addr,
        start_addr + 4,
        start_addr + user_size - 4,
        start_addr + user_size,
        start_addr + 0x344F4,
        start_addr + 0x4F750,
        start_addr + 0x6CE0C,
        start_addr + 0x8FEE8,
        start_addr + 0xB767C,
        start_addr + 0xD0A00,
        start_addr + 0xF11D0,
        start_addr + 0xC5E54,
        start_addr + 0xE9028,
        start_addr + 0x023BC,
        #read 
        start_addr,
        start_addr + 4,
        start_addr + user_size - 4,
        start_addr + user_size,
        start_addr + 0x344F4,
        start_addr + 0x4F750,
        start_addr + 0x6CE0C,
        start_addr + 0x8FEE8,
        start_addr + 0xB767C,
        start_addr + 0xD0A00,
        start_addr + 0xF11D0,
        start_addr + 0xC5E54,
        start_addr + 0xE9028,
        start_addr + 0x023BC,
        start_addr + user_size 
    )
    data_arr = (
        0x0314DFE1,
        0x3704E836,
        0x5208E431,
        0x77748E32,
        0x89158A64,
        0xAE603480,
        0xD24B086A,
        0xE50B1442,
        0xB23D7EFD,
        0x41C35871,
        0xC0E1638A,
        0x8E16CDA9,
        0x42EB0C85,
        0xBF9E7B2E, 
        #read
        0x0314DFE1,
        0x3704E836,
        0x5208E431,
        0x77748E32,
        0x89158A64,
        0xAE603480,
        0xD24B086A,
        0xE50B1442,
        0xB23D7EFD,
        0x41C35871,
        0xC0E1638A,
        0x8E16CDA9,
        0x42EB0C85,
        0xBF9E7B2E
    )
    await configure_userdesign(caravelEnv, used_addr=addr_arr)
    print([hex(i) for i in addr_arr])
    for addr, data in zip(addr_arr, data_arr):
        await RisingEdge(ack_hdl)
        await NextTimeStep()
        if addr_hdl.value.integer != addr:
            cocotb.log.error(
                f"[TEST] seeing unexpected address {hex(addr_hdl.value.integer)} expected {hex(addr)}"
            )
        elif we_hdl.value.integer == 1:# write
            if data_i_hdl.value.integer != data:
                cocotb.log.error(
                    f"[TEST] seeing unexpected write data {hex(data_i_hdl.value.integer)} expected {hex(data)} address {hex(addr)}"
                )
            else:
                cocotb.log.info(
                    f"[TEST] seeing the correct data {hex(data)} from address {hex(addr)}"
                )
        elif we_hdl.value.integer == 0:# read
            if data_o_hdl.value.integer != data:
                cocotb.log.error(
                    f"[TEST] seeing unexpected read data {hex(data_o_hdl.value.integer)} expected {hex(data)} address {hex(addr)}"
                )
        
            else:
                cocotb.log.info(
                    f"[TEST] seeing the correct data {hex(data)} from address {hex(addr)}"
                )
